### ISSUE TITLE
Info request times out with Orbstack

### ISSUE THREAD

**Issue by ciaranmcnulty** [beluga-php/docker-php#70]
I'm using [Orbstack](https://orbstack.dev) the majority of the time as a Docker Desktop for Mac replacement.

The test suite currently blocks on `tests/DockerTest.php` when reading the contents of the response to a request to `/v1.43/info`

Debugging through the HTTP client I can see an `\Http\Message\Encoding\DechunkStream` is being used for the response but for some reason although its buffer fills with the following, getContents() never returns:

```json
{"ID":"cd6466c2-a9c2-4b2b-835a-970306beb8d8","Containers":0,"ContainersRunning":0,"ContainersPaused":0,"ContainersStopped":0,"Images":13,"Driver":"overlay2","DriverStatus":[["Backing Filesystem","btrfs"],["Supports d_type","true"],["Using metacopy","false"],["Native Overlay Diff","true"],["userxattr","false"]],"Plugins":{"Volume":["local"],"Network":["bridge","host","ipvlan","macvlan","null","overlay"],"Authorization":null,"Log":["awslogs","fluentd","gcplogs","gelf","journald","json-file","local","logentries","splunk","syslog"]},"MemoryLimit":true,"SwapLimit":true,"CpuCfsPeriod":true,"CpuCfsQuota":true,"CPUShares":true,"CPUSet":true,"PidsLimit":true,"IPv4Forwarding":true,"BridgeNfIptables":true,"BridgeNfIp6tables":true,"Debug":false,"NFd":26,"OomKillDisable":false,"NGoroutines":43,"SystemTime":"2024-03-18T17:42:19.196730961Z","LoggingDriver":"json-file","CgroupDriver":"cgroupfs","CgroupVersion":"2","NEventsListener":1,"KernelVersion":"6.5.13-orbstack-00121-ge428743e4e98","OperatingSystem":"OrbStack","OSVersion":"","OSType":"linux","Architecture":"aarch64","IndexServerAddress":"https://index.docker.io/v1/","RegistryConfig":{"AllowNondistributableArtifactsCIDRs":null,"AllowNondistributableArtifactsHostnames":null,"InsecureRegistryCIDRs":["127.0.0.0/8"],"IndexConfigs":{"docker.io":{"Name":"docker.io","Mirrors":[],"Secure":true,"Official":true}},"Mirrors":null},"NCPU":8,"MemTotal":12526673920,"GenericResources":null,"DockerRootDir":"/var/lib/docker","HttpProxy":"","HttpsProxy":"","NoProxy":"","Name":"orbstack","Labels":[],"ExperimentalBuild":false,"ServerVersion":"24.0.7","Runtimes":{"io.containerd.runc.v2":{"path":"runc"},"runc":{"path":"runc"}},"DefaultRuntime":"runc","Swarm":{"NodeID":"","NodeAddr":"","LocalNodeState":"inactive","ControlAvailable":false,"Error":"","RemoteManagers":null},"LiveRestoreEnabled":false,"Isolation":"","InitBinary":"docker-init","ContainerdCommit":{"ID":"8c087663b0233f6e6e2f4515cee61d49f14746a8","Expected":"8c087663b0233f6e6e2f4515cee61d49f14746a8"},"RuncCommit":{"ID":"82f18fe0e44a59034f3e1f45e475fa5636e539aa","Expected":"82f18fe0e44a59034f3e1f45e475fa5636e539aa"},"InitCommit":{"ID":"de40ad0","Expected":"de40ad0"},"SecurityOptions":["name=seccomp,profile=builtin","name=cgroupns"],"ProductLicense":"Community Engine","DefaultAddressPools":[{"Base":"192.168.215.0/24","Size":24},{"Base":"192.168.228.0/24","Size":24},{"Base":"192.168.247.0/24","Size":24},{"Base":"192.168.207.0/24","Size":24},{"Base":"192.168.167.0/24","Size":24},{"Base":"192.168.107.0/24","Size":24},{"Base":"192.168.237.0/24","Size":24},{"Base":"192.168.148.0/24","Size":24},{"Base":"192.168.214.0/24","Size":24},{"Base":"192.168.165.0/24","Size":24},{"Base":"192.168.227.0/24","Size":24},{"Base":"192.168.181.0/24","Size":24},{"Base":"192.168.158.0/24","Size":24},{"Base":"192.168.117.0/24","Size":24},{"Base":"192.168.155.0/24","Size":24},{"Base":"192.168.147.0/24","Size":24},{"Base":"192.168.229.0/24","Size":24},{"Base":"192.168.183.0/24","Size":24},{"Base":"192.168.156.0/24","Size":24},{"Base":"192.168.97.0/24","Size":24},{"Base":"192.168.171.0/24","Size":24},{"Base":"192.168.186.0/24","Size":24},{"Base":"192.168.216.0/24","Size":24},{"Base":"192.168.242.0/24","Size":24},{"Base":"192.168.166.0/24","Size":24},{"Base":"192.168.239.0/24","Size":24},{"Base":"192.168.223.0/24","Size":24},{"Base":"192.168.164.0/24","Size":24},{"Base":"192.168.163.0/24","Size":24},{"Base":"192.168.172.0/24","Size":24},{"Base":"172.17.0.0/16","Size":16},{"Base":"172.18.0.0/16","Size":16},{"Base":"172.19.0.0/16","Size":16},{"Base":"172.20.0.0/14","Size":16},{"Base":"172.24.0.0/14","Size":16},{"Base":"172.28.0.0/14","Size":16}],"Warnings":null}
```

It's possible Orbstack is behaving wrongly here (maybe never EOFing?) but I'm not sure how to proceed debugging the socket connection.

**Comment by ciaranmcnulty** [comment]
`curl --unix-socket /var/run/docker.sock http://localhost/v1.43/info`

If I run this command I get the same data and the terminal returns (no blocking)

**Comment by flavioheleno** [comment]
hey @ciaranmcnulty, thank you for pointing this out!

I'll try to reproduce it locally, do you mind sending over the version of your macOS, Orbstack and anything else you may find could be related?

cheers!

**Comment by ciaranmcnulty** [comment]
I'm using MacOS Sonoma (14.0) and Orbstack 1.4.3

If you try it out please be aware that Orbstack will symlink `/var/run/docker.sock` to its own socket so it contends with Docker Desktop somewhat

**Comment by ciaranmcnulty** [comment]
Some debugging, what seems to happen is that the raw stream containing the `/v1.43/info` is read, and is then in the following exhausted state (from `stream_get_metadata()`:

```php
array (
  'timed_out' => false,
  'blocked' => true,
  'eof' => false,
  'stream_type' => 'unix_socket',
  'mode' => 'r+',
  'unread_bytes' => 0,
  'seekable' => false,
)
```

Then the HTTP client does another `fread` (for I'm sure a good reason) which then blocks for the default socket timeout. If I set that low I can observe everything breaking with `Failed to read from stream`

My instinct is that Orbstack might be doing something wrong here by not setting the stream as EOF, but my knowledge of how HTTP-over-local-socket is supposed to work is letting me down for further debugging!

**Comment by ciaranmcnulty** [comment]
Here is a minimal PHP script that times out when reading from Orbstack

Unless there is a socket stream option we can tweak here to change the behaviour it may be something we need to report upstream.

It's odd that Curl and the Docker CLI tools don't have an issue though.

```
use Http\Client\Common\Plugin\AddHostPlugin;
use Http\Client\Common\PluginClientFactory;
use Http\Discovery\Psr17FactoryDiscovery;

include __DIR__ . '/vendor/autoload.php';

$client = (new PluginClientFactory())->createClient(
    new Http\Client\Socket\Client(
        [
//            'remote_socket' => 'unix:///Users/ciaranmcnulty/.orbstack/run/docker.sock',
            'remote_socket' => 'unix:///Users/ciaranmcnulty/.docker/run/docker.sock',
        ]
    ),
    [
        new AddHostPlugin(Psr17FactoryDiscovery::findUriFactory()->createUri('http://localhost')),
    ],
);

$request = Psr17FactoryDiscovery::findRequestFactory()->createRequest('GET', '/v1.43/info');
$response = $client->sendRequest($request);

var_dump($response->getBody()->getContents());
```

**Comment by flavioheleno** [comment]
hey @ciaranmcnulty, sorry that I haven't got back to this so far, things are a bit busy on my end.

I've just tested your script and it also works fine with `podman`. I think the issue is on Orbstack as you guessed..

**Comment by ciaranmcnulty** [comment]
After some more investigation it seems that the HTTP library being used adds a "Connection: close" header when using a socket, but Orbstack is not respecting it.

I've reported there https://github.com/orbstack/orbstack/issues/1119 so this issue could possibly be closed.

**Comment by flavioheleno** [comment]
I'll take a look at other available HTTP Client options

**Comment by ciaranmcnulty** [comment]
This is apparently fixed upstream so once 1.5.2 I'll test and close this issue

**Comment by ciaranmcnulty** [comment]
Orbstack 1.6.0 Canary 1 fixes this
